YAML Model for Semantic Conventions

thisthat opened the pull request:
**Ready for merge**

# YAML model for Semantic Convention 

This is a proposal to define semantic conventions in a YAML format that resolves #547.

The different language implementations can use `git submodule`s to reference these definitions, so the following can be automatically generated: 

- Constants for semantic attributes  

- Typed Spans (type-safe span builders with named methods/functions rather than having users type or copy-paste strings from the spec doc for attribute keys) 

 For the `opentelemetry-specification` repo, markdown tables will be generated from the YAML files to serve as written documentation. 

## Markdown Tables 

Semantic conventions use Markdown tables to describe the list of attributes that belong to them. 

With a common format, all tables for the semantic conventions will have the same appearance and requirements. It is expected that special HTML tags are used in skeleton/template Markdown files to tag where and which table must be generated, as is with [markdown-toc](https://github.com/jonschlinkert/markdown-toc) that is currently used to generate the TOCs. 

  

  

## Constants for Semantic Attributes 

Currently, for each instrumentation, attribute names must be copied manually from the specification, which is an error-prone activity. Using a common YAML format, languages can automatically generate constant keys for Semantic Attributes, e.g. [JS](https://github.com/open-telemetry/opentelemetry-js/blob/3d9b8228b536c99e09d36df7ac3386e3239f067a/packages/opentelemetry-plugin-http/src/enums/AttributeNames.ts), and [Java](https://github.com/open-telemetry/opentelemetry-java/blob/master/api/src/main/java/io/opentelemetry/trace/attributes/SemanticAttributes.java). 

  

  

## Typed Spans 

  

Semantic Attributes are an important aspect of OpenTelemetry because they allow to precisely define the information that provides insights into a specific part of a system. Defining the data format using semantic conventions is the only reliable way to allow back ends to reason about the telemetry data they receive and to analyze them further. To support OpenTelemetry users in properly supplying this data, we can provide additional APIs that aid building spans for specific operations by correctly setting the semantic attributes for a given semantic convention. 

For more information, please refer to the prior art in [OpenCensus](https://github.com/census-instrumentation/opencensus-specs/issues/257), [OTEP25](https://github.com/open-telemetry/oteps/pull/25/files), [original specifications](https://github.com/open-telemetry/opentelemetry-specification/pull/14), and the [Java proposal](https://github.com/open-telemetry/opentelemetry-java/pull/964/files). 

Note that, each implementation will define its own API structure for the Typed Spans to use the common patterns offered by the language. 

  

# Description of the model 

  

The fields and their expected values are presented in [allowed_syntax.md](https://github.com/dynatrace-oss-contrib/opentelemetry-specification/blob/semconv-model/yaml/allowed_syntax.md). 

  

For a basic attribute description, please refer to the [General Attributes example](https://github.com/dynatrace-oss-contrib/opentelemetry-specification/blob/semconv-model/yaml/span_attribute/general.yaml). 

Here, each attribute is specified in a list providing information for the key, which is computed by joining the semantic convention prefix and the attribute name. Furthermore, the type expected for each attribute is defined. In some cases, attributes are relevant to perform sampling decisions. These attributes can be marked as `sampling_relevant`, for example in the gRPC and FaaS semantic conventions the attributes [service](https://github.com/dynatrace-oss-contrib/opentelemetry-specification/blob/semconv-model/yaml/span_attribute/rpc.yaml#L9) and [trigger](https://github.com/dynatrace-oss-contrib/opentelemetry-specification/blob/semconv-model/yaml/span_attribute/faas.yaml#L11), respectively. Marking an attribute as `sampling_relevant` will help the Typed Spans to request such information before the Span is started so it is available via attributes to the [Sampler](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#sampler). Furthermore, generated Markdown tables will also have a column to indicate if an attribute is sampling relevant or not. 

In the case that an enumeration is expected, the pre-defined enum values are listed and we can additionally set the flag `allow_custom_values` to true to allow other, custom values to be provided. This will help the Typed Span APIs to properly restrict the allowed attribute values while building the span. 

  

In some semantic conventions, a common structure is shared by specialized semantic conventions for specific cases, e.g. [HTTP](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/http.md) and [gRPC](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/rpc.md), where a common description is given and then additional attributes are described depending on whether it is a server or client operation. To express this, the `extends` field is added and set to the id of the semantic convention to extend, e.g [gRPC Client](https://github.com/dynatrace-oss-contrib/opentelemetry-specification/blob/semconv-model/yaml/span_attribute/rpc.yaml#L17). 

  

Another common case is to reference attributes of another semantic convention. For example, the [gRPC](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/rpc.md) semantic convention references many [Network Attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/span-general.md#general-network-connection-attributes). To cross-reference attributes between different semantic conventions the `ref` field is used. `ref` also allows overriding the description of the attributes to properly fit it to the current use case. When using `ref`, the constraints that specify if the referenced attribute is required are not taken over, this always has to be explicitly defined for each convention in which it is used. 

An example is the [gRPC Client](https://github.com/dynatrace-oss-contrib/opentelemetry-specification/blob/semconv-model/yaml/span_attribute/rpc.yaml#L21). Using `ref` will allow the Markdown table generation to create links between the different semantic conventions. 

  

Semantic Conventions might have special constraints, for example in [gRPC](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/rpc.md) at least one of `net.peer.ip` or `net.peer.port` must be set. This can be expressed in the YAML format using the `any_of` constraint. An example is the [RPC Semantic Convention](https://github.com/dynatrace-oss-contrib/opentelemetry-specification/blob/semconv-model/yaml/span_attribute/rpc.yaml#L13). 

Another constraint is that a semantic convention requires that another semantic convention is, in turn, followed as well. For example, in the [FaaS Semantic Convention](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/faas.md#http) when a function is used to respond an inbound HTTP request, the [HTTP Server Semantic Convention](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/http.md#http-server-semantic-conventions) must also be followed. This can be expressed in the YAML format using the `include` constraint. An example is available [here](https://github.com/dynatrace-oss-contrib/opentelemetry-specification/blob/semconv-model/yaml/span_attribute/faas.yaml#L82). 

# Versioning

Since OpenTelemetry has reached the beta status, it is important to introduce the notion of versioning. Whenever we introduce a semantic change, the version should be bumped by one. This way, backends can correctly ingest OpenTelemetry data.
What we consider a semantic change, is out of the scope of this PR and should be discussed in a separate issue.

bogdandrutu commented:
I think this is a very nice PR and well documented, but would like to rise some concerns:
* In this PR we are adding some new properties to the attributes such as `sampling_relevant` and `allow_custom_values`. I would prefer these to be added later as enhancements to the current model.
* We are talking about `Typed Spans` which has not been defined yet.

So I would focus for the moment on not adding new functionality.

bogdandrutu commented on yaml/span_attribute/faas.yaml:
```diff
@@ -0,0 +1,108 @@
+          For example, in Cloud Storage or S3 corresponds to the bucket name,
+          and in Cosmos DB to the database name.
+        examples: ['myBucketName', 'myDbName']
+      - id: operation
+        required: always
+        type:
+          allow_custom_values: true
+          members:
+            - id: insert
+              value: 'insert'
+            - id: edit
+              value: 'edit'
+            - id: delete
+              value: 'delete'
+        brief: 'Describes the type of the operation that was performed on the data.'
```
This is not defined right now, maybe consider to remove it, we should not define new semantic conventions here.

Oberon00 replied:
What do you mean? I think this is copied straight from https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/faas.md#datasource

Oberon00 commented on yaml/allowed_syntax.md:
```diff
@@ -0,0 +1,238 @@
+# extends MUST point to an existing semconv id
+extends ::= string
+
+span_kind ::= "client"
```
However, semantic conventions define which span kind you need to set for certain Spans. E.g. PRODUCER for most messaging senders, SERVER for an HTTP server, etc.

bogdandrutu replied:
My comment was, do we need to define the values of the SpanKind here ?

Oberon00 commented on yaml/allowed_syntax.md:
```diff
@@ -0,0 +1,238 @@
+ref ::= id
+
+type ::= "string"
+     |   "number"
```
This is unfortunately still an unresolved discussion, currently we do indeed have only vaguely defined "numeric" attributes, see #425.

bogdandrutu replied:
I am +1 to update specs to say explicitly what numeric values we accept

Oberon00 commented on yaml/allowed_syntax.md:
```diff
@@ -0,0 +1,238 @@
+     |   "string[]"
+     |   "number[]"
+     |   "boolean[]"
+     |   enum
```
Enums are just the YAML's way of saying that only certain string/number values are allowed or to provide a list of constants that are recommended for them. This does not propose a new attribute type.

bogdandrutu replied:
Document that. Also what is the difference between `enum` and `allow_custom_values`

thisthat commented on yaml/allowed_syntax.md:
```diff
@@ -0,0 +1,238 @@
+required ::= "yes"
+         |   "conditional"
+
+sampling_relevant ::= boolean
```
I agree that this is a new concept. I will remove it and propose it as a new concept in a separate issue.

bogdandrutu replied:
Please remove.

thisthat commented:
> I think this is a very nice PR and well documented, but would like to rise some concerns:
> 
> * In this PR we are adding some new properties to the attributes such as `sampling_relevant` and `allow_custom_values`. I would prefer these to be added later as enhancements to the current model.
> * We are talking about `Typed Spans` which has not been defined yet.
> 
> So I would focus for the moment on not adding new functionality.

I do agree with you @bogdandrutu that `sampling_relevant` is a new concept that we should not introduce here. However, `allow_cutom_values` it is not necessarily new. I would argue that currently, it is an implicitly concept (see [`db.type`](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/database.md), [`faas.trigger`](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/faas.md), [`http.flavor`](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/http.md), [`messaging.system`](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/messaging.md), etc) that now is explicit and properly documented.

Regarding your second point, this PR proposes the vision of having Typed Spans. I would argue that this is a (side-effect) benefit we can have for free using the proposed YAML model and not a concept that we are introducing with this PR.

bogdandrutu commented:
> Regarding your second point, this PR proposes the vision of having Typed Spans. I would argue that this is a (side-effect) benefit we can have for free using the proposed YAML model and not a concept that we are introducing with this PR.

Based on the description it seems like the main reason to do this.

thisthat commented:
@open-telemetry/specs-approvers @open-telemetry/specs-trace-approvers please take another look. 
I add a small example of how we could model with YAML files network and HTTP semantic conventions. I am using these two semantic conventions to showcase how markdown tables can be automatically generated from such definitions. Similarly, I am proposing typed spans in the java-instrumentation repository that are automatically generated from these definitions (https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/502).

In this PR, there is also the source code used to generate markdown tables and code. It is not perfect and it requires some more polishing, but we are using it for quite a while internally at Dynatrace. It can be used as a first draft to start avoiding to manually replace tables and code after every semantic convention change (e.g. https://github.com/heyams/opentelemetry-java/blob/master/api/src/main/java/io/opentelemetry/trace/attributes/SemanticAttributes.java).
The idea is to donate this code to the OpenTelemetry project so everyone can use and improve it.

arminru commented:
The semantic convention generator which makes use of the YAML model proposed here is available for review in the new build-tools repo: https://github.com/open-telemetry/build-tools/pull/5

SergeyKanzhelev commented:
@arminru @thisthat It would be great to see this PR adjusted accordingly to use the external tool to minimize the number of files to review.

I'd also suggest to minimize the amount of semantics changes while introducing yaml for semantic conventions. I'd really love to have a yaml declarations merged so we can iterate on it.

thisthat commented:
@SergeyKanzhelev I agree. My plan is to split in multiple PRs the transition to YAML declarations. I will update this PR proposing only a single semantic convention so we can start iterate on it :)

thisthat commented:
I updated the PR proposing the semantic convention for http and general attributes. 
I added a Makefile entry to run the docker image that generates the semantic convention tables using the yaml definitions. Furthermore, I created a GitHub action that uses the docker image to verify that the tables are up-to-date with the yaml definitions.
Since https://github.com/open-telemetry/build-tools/pull/5 is not yet approved, I have uploaded the docker image to my personal account and use it in this PR. As soon as the image is published under the OTel account, I will update the required files.

Oberon00 commented on yaml/allowed_syntax.md:
```diff
@@ -0,0 +1,235 @@
+# Semantic Convention YAML Language
```
Rename to just syntax.md?

SergeyKanzhelev replied:
Also a better location would be somewhere under semantic_convention I think? Or name the folder samantic_convention?

Oberon00 commented on specification/trace/semantic_conventions/http.md:
```diff
@@ -189,24 +201,24 @@ If the route does not include the application root, it SHOULD be prepended to th
-* `http.scheme`, `http.server_name`, `net.host.port`, `http.target`
-* `http.scheme`, `net.host.name`, `net.host.port`, `http.target`
+* `http.scheme`, `http.server_name`, [net.host.port](span-general.md), `http.target`
+* `http.scheme`, [net.host.name](span-general.md), [net.host.port](span-general.md), `http.target`
```
It seems we lost the `code` markup here. The tool should be updated to generate

```suggestion
* `http.scheme`, [`net.host.name`](span-general.md), [`net.host.port`](span-general.md), `http.target`
```

thisthat replied:
Yes, I would like to have an agreement on style and wording to use. So I can update the code of the tool to support it.

SergeyKanzhelev replied:
easiest is to try to make a tool generate something that we used to have here and then update it in a separate PR. Least controversial approach for sure.

Oberon00 commented on specification/trace/semantic_conventions/http.md:
```diff
@@ -104,12 +113,15 @@ For an HTTP client span, `SpanKind` MUST be `Client`.
-One of the following sets of attributes is required (in order of usual preference unless for a particular web client/framework it is known that some other set is preferable for some reason; all strings must be non-empty):
+<!-- semconv http.client -->
+
+At least one of the following is required:
```
```suggestion
**Additional attribute requirements:** At least one of the following sets of attributes is required:
```

etc

Oberon00 commented on Makefile:
```diff
@@ -33,3 +35,8 @@ install-markdown-lint:
+
+.PHONY: table-generation
+table-generation:
+	docker run --rm -v $(PWD)/yaml:/source -v $(PWD)/specification:/spec thisthatdc/semconvgen -f /source markdown -md /spec
```
I hope this is configured to fail on mismatch? If so, please add a comment and rename the target. I don't think changing the tables from a github action is really possible. The action would need to have commit rights into every PR's repository.

thisthat replied:
This is a Makefile, not the GitHub action. The action only checks if the tables are up-to-date.

Oberon00 commented:
We use this YAML already for resources at Dynatrace without issue. I assume metric label(sets) should work just as well, although the tool has no special support for metrics.

SergeyKanzhelev commented on .github/workflows/semantic-convention-check.yaml:
```diff
@@ -0,0 +1,18 @@
+    steps:
+    - uses: actions/checkout@v1
+    - name: Verify that semantic convention tables are up-to-date
+      run: docker run --rm -v $(pwd)/yaml:/source -v $(pwd)/specification:/spec thisthatdc/semconvgen -f /source markdown -md /spec --md-check
```
what is `thisthatdc` here?

thisthat replied:
That's my private account. As I said in my comment, I will update to `otel` as soon as the other PR gets merged.

SergeyKanzhelev commented on .github/workflows/semantic-convention-check.yaml:
```diff
@@ -0,0 +1,18 @@
+    branches: [ master ]
+    paths:
+      - .github/workflows/semantic-convention-check.yml
+      - 'spec/**'
```
should it be limited to semantic convention files?

thisthat commented on Makefile:
```diff
@@ -33,3 +35,8 @@ install-markdown-lint:
+
+.PHONY: table-generation
+table-generation:
+	docker run --rm -v $(PWD)/yaml:/source -v $(PWD)/specification:/spec thisthatdc/semconvgen -f /source markdown -md /spec
```
I will leave this under my personal account for the moment. This way, people can experiment and try the tool. As soon as the tool is published with the opentelemety account, I will update the make file.

Oberon00 reviewed (approved):
LGTM, but before merging this, the build needs to become green:
* Either https://github.com/open-telemetry/build-tools/pull/5 needs to be merged first, or the GH action needs to be disabled/commented out for now.
* The docfx check needs to be fixed (the `nettransport-attribute` in `span-general.md` anchor is missing):

```
[20-08-18 11:34:37.018]Warning:[Postprocess.HandlePostProcessorsWithIncremental.HandlePostProcessors.Processing html.ValidateBookmark](specification/trace/semantic_conventions/span-general.md#L12)Invalid link: '[`net.transport` attribute](#nettransport-attribute)'. The file specification/trace/semantic_conventions/span-general.md doesn't contain a bookmark named 'nettransport-attribute'.
[20-08-18 11:34:37.018]Warning:[Postprocess.HandlePostProcessorsWithIncremental.HandlePostProcessors.Processing html.ValidateBookmark](specification/trace/semantic_conventions/rpc.md#L64)Invalid link: '[net.transport][]'. The file specification/trace/semantic_conventions/span-general.md doesn't contain a bookmark named 'nettransport-attribute'.
[20-08-18 11:34:37.018]Warning:[Postprocess.HandlePostProcessorsWithIncremental.HandlePostProcessors.Processing html.ValidateBookmark](specification/trace/semantic_conventions/messaging.md#L126)Invalid link: '[`net.transport`][]'. The file specification/trace/semantic_conventions/span-general.md doesn't contain a bookmark named 'nettransport-attribute'.
[20-08-18 11:34:37.018]Warning:[Postprocess.HandlePostProcessorsWithIncremental.HandlePostProcessors.Processing html.ValidateBookmark](specification/trace/semantic_conventions/database.md#L42)Invalid link: '[`net.transport`][]'. The file specification/trace/semantic_conventions/span-general.md doesn't contain a bookmark named 'nettransport-attribute'.
	4 Warning(s)
```

arminru commented:
Still pending on https://github.com/open-telemetry/build-tools/pull/5 to be merged (or the build action to be disabled for now).

bogdandrutu commented:
Rebase?

arminru commented on Makefile:
```diff
@@ -34,3 +36,7 @@ install-markdown-lint:
+
+.PHONY: table-generation
+table-generation:
+	docker run --rm -v $(PWD)/semantic_conventions:/source -v $(PWD)/specification:/spec thisthatdc/semconvgen -f /source markdown -md /spec
```
```suggestion
	docker run --rm -v $(PWD)/semantic_conventions:/source -v $(PWD)/specification:/spec otel/semconvgen -f /source markdown -md /spec
```
Should be working now that https://github.com/open-telemetry/build-tools/pull/5 was merged and the image published to https://hub.docker.com/r/otel/semconvgen.

SergeyKanzhelev commented:
it shows me this strange UI without Update Branch or Merge button:

![image](https://user-images.githubusercontent.com/9950081/91481844-e92d1280-e859-11ea-87c3-14684b2bb319.png)

@bogdandrutu does it show the same for you?

arminru commented:
@SergeyKanzhelev 
@bogdandrutu enabled a setting that all branches have to be brought up to date manually before merging until all PRs picked up some linting fixes he recently added (see #512). I updated the branch manually just now.

SergeyKanzhelev commented:
> @SergeyKanzhelev
> @bogdandrutu enabled a setting that all branches have to be brought up to date manually before merging until all PRs picked up some linting fixes he recently added (see #512). I updated the branch manually just now.

Understood. On other PRs I see a button to update a branch so I can do it myself:

![image](https://user-images.githubusercontent.com/9950081/91490877-023cc000-e868-11ea-9762-558b9ac4fb80.png)

Maybe this is because I don't have write permissions on DynaTrace repo to update the branch myself before merging.
